### PR TITLE
#20 delivery log domain

### PR DESCRIPTION
**Pull request by PerikK**

### CODE REVIEW

File: src/controllers/deliveryLog.js [hunk 1]
```diff
@@ -1,24 +1,58 @@
...
+    return sendDataResponse(res, 201, {
+      data: {
+        log: {
+          id: log.id,
+          cohortId: cohortId,
+          date: log.date,
+          author: {
+            id: log.user.id,
+            firstName: req.user.firstName,
+            lastName: req.user.lastName
```
Comment (tzoltie) [inline src/controllers/deliveryLog.js]: user could be destructured from the request to make these lines cleaner

File: src/controllers/deliveryLog.js [hunk 2]
```diff
@@ -1,24 +1,58 @@
...
-          id: index + 1,
-          content: line.content
+  try {
+    if (!cohortId || !lines) {
+      throw Error(ERR.INCOMPLETE_REQUEST)
+    }
+
+    const existing = await getCohortById(cohortId)
+    if (!existing) {
+      throw Error(ERR.COHORT_NOT_FOUND)
```
Comment (angustownsley) [inline src/controllers/deliveryLog.js]: I would prefer this be 
```js
const cohortExists = await getCohortById(cohortId)
    if (!cohortExists )
```
to make it clear what this is checking.

File: src/controllers/deliveryLog.js [hunk 3]
```diff
@@ -1,24 +1,58 @@
...
+      })
+    } else if (error.message === ERR.COHORT_NOT_FOUND) {
+      return res.status(404).json({
+        status: 'error',
+        data: error.message
+      })
+    } else {
+      return res.status(500).json({
+        status: 'error',
+        message: 'An error occurred while creating the delivery log.'
```
Comment (angustownsley) [inline src/controllers/deliveryLog.js]: This message could be moved to errors.js

**Review by angustownsley — CHANGES_REQUESTED** [review]

**Review by angustownsley — DISMISSED** [review]

**Review by homonoviscoding — DISMISSED** [review]

File: src/controllers/deliveryLog.js [hunk 4]
```diff
@@ -1,24 +1,59 @@
...
-        first_name: req.user.firstName,
-        last_name: req.user.lastName
-      },
-      lines: lines.map((line, index) => {
-        return {
-          id: index + 1,
-          content: line.content
+  try {
+    if (!cohortId || !lines) {
+      throw Error(ERR.INCOMPLETE_REQUEST)
```
Comment (Auenc) [inline src/controllers/deliveryLog.js]: Why are we throwing an error only to handle it in the catch? Couldn't we just check if the `cohortId` and `lines` exists prior to the try block and just return the response?

File: src/controllers/deliveryLog.js [hunk 5]
```diff
@@ -1,24 +1,59 @@
...
+        log: {
+          id: log.id,
+          cohortId: cohortId,
+          date: log.date,
+          author: {
+            id: log.user.id,
+            firstName,
+            lastName
+          },
+          lines: log.lines.map((line) => ({
```
Comment (Auenc) [inline src/controllers/deliveryLog.js]: let's do this map prior to the return

File: src/controllers/deliveryLog.js [hunk 6]
```diff
@@ -1,24 +1,59 @@
...
+            id: line.id,
+            content: line.content
+          }))
         }
+      }
+    })
+  } catch (error) {
+    console.error(error)
```
Comment (Auenc) [inline src/controllers/deliveryLog.js]: Add some context to this error message

File: src/controllers/deliveryLog.js [hunk 7]
```diff
@@ -1,24 +1,59 @@
...
+      return res.status(400).json({
+        status: 'error',
+        data: error.message
+      })
+    } else if (error.message === ERR.COHORT_NOT_FOUND) {
+      return res.status(404).json({
+        status: 'error',
+        data: error.message
+      })
+    } else {
```
Comment (Auenc) [inline src/controllers/deliveryLog.js]: redundant else, just return the 500

File: src/domain/cohort.js [hunk 8]
```diff
@@ -25,3 +25,12 @@ export class Cohort {
     }
   }
 }
+
+export const getCohortById = async (cohortId) => {
```
Comment (Auenc) [inline src/domain/cohort.js]: this could be a one line function. We can remove async too.

File: src/domain/deliveryLog.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  return await dbClient.deliveryLog.create({
+    data: {
+      cohort: {
+        connect: { id: cohortId }
+      },
+      user: {
+        connect: { id: userId }
+      },
+      lines: {
+        create: linesArray.map((line) => ({
```
Comment (Auenc) [inline src/domain/deliveryLog.js]: let's do the map before

File: src/domain/deliveryLog.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import dbClient from '../utils/dbClient.js'
+
+export const createDeliveryLogDb = async (cohortId, userId, lines) => {
+  const linesArray = Array.isArray(lines) ? lines : []
+
+  return await dbClient.deliveryLog.create({
```
Comment (Auenc) [inline src/domain/deliveryLog.js]: await is redundant here

**Review by Auenc — CHANGES_REQUESTED** [review]

File: src/controllers/deliveryLog.js [hunk 11]
```diff
@@ -1,24 +1,55 @@
...
-          id: index + 1,
-          content: line.content
+  if (!cohortId || !lines) {
+    return res.status(400).json({
+      status: 'error',
+      data: ERR.INCOMPLETE_REQUEST
+    })
+  }
+
+  const cohortExists = await getCohortById(cohortId)
```
Comment (Auenc) [inline src/controllers/deliveryLog.js]: this needs to be in the try catch because it can error

**Review by Auenc — CHANGES_REQUESTED** [review]

File: src/controllers/deliveryLog.js (hunk 6 above)
Comment (PerikK) [reply src/controllers/deliveryLog.js]: What you mean by this? It's a generic error. Server died or something. :D

Comment (Auenc) [reply src/controllers/deliveryLog.js]: Something like `console.error('error creating delivery log', error)`

**Review by Auenc — APPROVED** [review]

**Review by angustownsley — APPROVED** [review]

File: .env.example [hunk 12]
```diff
@@ -1,4 +0,0 @@
-PORT=4000
```
Comment (Luca-Terrazzan) [inline .env.example:1]: This shouldn't be deleted :cry:

**Pull request merged**